meadow-client/src/MonadApp.purs

```diff
@@ -0,0 +1,394 @@
+  HalogenApp m (WebData a)
+runAjax action = wrap $ RemoteData.fromEither <$> runExceptT action
+
+-- | Handles the messy business of running an editor command if the
+-- editor is up and running.
+withEditor ::
+  forall m a.
+  MonadEffect m =>
+  (Editor -> Effect a) ->
+  HalogenApp m (Maybe a)
+withEditor action =
+  HalogenApp
+    $ do
+        mEditor <- query' cpEditor EditorSlot $ request GetEditor
+        case join mEditor of
+          Just editor -> Just <$> (liftEffect $ action editor)
+          _ -> pure Nothing
+
+withMarloweEditor ::
+  forall m a.
+  MonadEffect m =>
+  (Editor -> Effect a) ->
+  HalogenApp m (Maybe a)
+withMarloweEditor action =
+  HalogenApp
+    $ do
+        mEditor <- query' cpMarloweEditor MarloweEditorSlot $ request GetEditor
+        case mEditor of
+          Just (Just editor) -> do
+            liftEffect $ Just <$> action editor
+          _ -> pure Nothing
+
+updateContractInStateP :: String -> MarloweState -> MarloweState
+updateContractInStateP text state = set _contract con state
+  where
+  con = case runParser text contract of
+    Right pcon -> Just pcon
+    Left _ -> Nothing
+
+updateStateP :: MarloweState -> MarloweState
+updateStateP oldState = actState
+  where
+  sigState = updateSignatures oldState
+
+  mSimulatedState = simulateState sigState
+
+  actState = case mSimulatedState of
+    Just simulatedState -> updateActions sigState simulatedState
+    Nothing -> sigState
+
+updateSignatures :: MarloweState -> MarloweState
+updateSignatures oldState = case oldState.contract of
+  Just oldContract -> over (_transaction <<< _signatures) (resizeSigs (peopleFromStateAndContract (oldState.state) oldContract)) oldState
+  Nothing -> oldState
+
+simulateState :: MarloweState -> Maybe {state :: State, contract :: Contract, outcome :: TransactionOutcomes, validity :: TransactionValidity}
+simulateState state = case state.contract of
+  Just c ->
+    Just
+      ( case applyTransaction inps sigs bn st c mic of
+        MSuccessfullyApplied {state: newState, contract: newContract, outcome: outcome} inputWarnings ->
+          { state: newState
+          , contract: newContract
+          , outcome: outcome
+          , validity: ValidTransaction inputWarnings
+          }
+        MCouldNotApply InvalidInput -> if (inps == Nil)
+          then
+            { state: st
+            , contract: reduce state.blockNum state.state c
+            , outcome: Map.empty
+            , validity: EmptyTransaction
+            }
+          else
+            { state: emptyState
+            , contract: Null
+            , outcome: Map.empty
+            , validity: InvalidTransaction InvalidInput
+            }
+        MCouldNotApply err ->
+          { state: emptyState
+          , contract: Null
+          , outcome: Map.empty
+          , validity: InvalidTransaction err
+          }
+      )
+  Nothing -> Nothing
```

Comment:
I'm realising this isn't your code, you've just reformatted it. But I can't-not mention the fact `case x of Just v -> Just (f v), Nothing -> Nothing` is just `map f x`.

Reply:
done with `mapFlipped`